src/cwe_checker_lib/src/analysis/taint/mod.rs

```diff
@@ -0,0 +1,521 @@
+    ///
+    /// # Default
+    ///
+    /// Just returns the proposed state.
+    fn update_def_post(
+        &self,
+        _old_state: &State,
+        new_state: State,
+        _def: &Term<Def>,
+    ) -> Option<State> {
+        Some(new_state)
+    }
+}
+
+impl<'a, T: TaintAnalysis<'a>> forward_interprocedural_fixpoint::Context<'a> for T {
+    type Value = State;
+
+    fn get_graph(&self) -> &Cfg<'a> {
+        self.get_cfg()
+    }
+
+    fn merge(&self, state1: &Self::Value, state2: &Self::Value) -> Self::Value {
+        state1.merge(state2)
+    }
+
+    fn specialize_conditional(
+        &self,
+        state: &Self::Value,
+        _condition: &Expression,
+        _block_before_condition: &Term<Blk>,
+        _is_true: bool,
+    ) -> Option<Self::Value> {
+        Some(state.clone())
+    }
+
+    fn update_call(
+        &self,
+        state: &Self::Value,
+        call: &Term<Jmp>,
+        target: &CfgNode,
+        calling_convention: &Option<String>,
+    ) -> Option<Self::Value> {
+        <Self as TaintAnalysis>::update_call(self, state, call, target, calling_convention)
+    }
+
+    fn update_call_stub(&self, state: &Self::Value, call: &Term<Jmp>) -> Option<Self::Value> {
+        <Self as TaintAnalysis>::update_call_stub(self, state, call)
+    }
+
+    fn update_jump(
+        &self,
+        state: &Self::Value,
+        jump: &Term<Jmp>,
+        untaken_conditional: Option<&Term<Jmp>>,
+        target: &Term<Blk>,
+    ) -> Option<Self::Value> {
+        <Self as TaintAnalysis>::update_jump(self, state, jump, untaken_conditional, target)
+    }
+
+    fn update_def(&self, state: &Self::Value, def: &Term<Def>) -> Option<Self::Value> {
+        if state.is_empty() {
+            // Without taint there is nothing to propagate.
+            return None;
+        }
+
+        let new_state = match &def.term {
+            Def::Assign { var, value } => self.update_def_assign(state, &def.tid, var, value),
+            Def::Load { var, address } => self.update_def_load(state, &def.tid, var, address),
+            Def::Store { address, value } => self.update_def_store(state, &def.tid, address, value),
+        };
+
+        self.update_def_post(state, new_state, def)
+    }
+
+    fn update_return(
+        &self,
+        state_before_return: Option<&State>,
+        state_before_call: Option<&State>,
+        call_term: &Term<Jmp>,
+        return_term: &Term<Jmp>,
+        calling_convention: &Option<String>,
+    ) -> Option<State> {
+        if let Some(state) = state_before_return {
+            <Self as TaintAnalysis>::update_return(
+                self,
+                state,
+                call_term,
+                return_term,
+                calling_convention,
+            )
+        }
+
+        if let Some(state) = state_before_call {
+            self.update_call_generic(state, &call_term.tid, calling_convention)
+        } else {
+            None
+        }
+    }
+}
```

Comment:
Do you want to give users the same control over the interprocedural data flow propagation as in the fixpoint traits? Then doing that in `update_call_generic` would probably still lead to confusion, because one would expect the same interface also having the same name between the fixpoint traits and taint analysis traits.

If you want to hide the details on how interprocedural taint propagation works (like in your current interface), then letting `TaintAnalysis::update_return` return an `Option<State>` is the way to go, but a naive merge will not be enough in the `update_return` implementation of the fixpoint trait. If we decide to go this way, I would be OK with just adding a `FIXME` note here and postponing the actual implementation, because your PR is already quite large.